signature-aggregator/aggregator/aggregator.go

```
@@ -0,0 +1,507 @@
+	s.mu.RLock()
+	subnetID, ok := s.subnetIDsByBlockchainID[blockchainID]
+	s.mu.RUnlock()
+	if ok {
+		return subnetID, nil
+	}
+	s.logger.Info("Signing subnet not found, requesting from PChain", zap.String("chainID", blockchainID.String()))
+	subnetID, err := s.network.GetSubnetID(blockchainID)
+	if err != nil {
+		return ids.ID{}, fmt.Errorf("source blockchain not found for chain ID %s", blockchainID)
+	}
+	s.SetSubnetID(blockchainID, subnetID)
+	return subnetID, nil
+}
+
+func (s *SignatureAggregator) SetSubnetID(blockchainID ids.ID, subnetID ids.ID) {
+	s.mu.Lock()
+	s.subnetIDsByBlockchainID[blockchainID] = subnetID
+	s.mu.Unlock()
+}
+
+// Attempts to create a signed warp message from the accumulated responses.
+// Returns a non-nil Warp message if [accumulatedSignatureWeight] exceeds the signature verification threshold.
+// Returns false in the second return parameter if the app response is not relevant to the current signature
+// aggregation request. Returns an error only if a non-recoverable error occurs, otherwise returns a nil error
+// to continue processing responses.
+func (s *SignatureAggregator) handleResponse(
+	response message.InboundMessage,
+	sentTo set.Set[ids.NodeID],
+	requestID uint32,
+	connectedValidators *peers.ConnectedCanonicalValidators,
+	unsignedMessage *avalancheWarp.UnsignedMessage,
+	signatureMap map[int]blsSignatureBuf,
+	accumulatedSignatureWeight *big.Int,
+	quorumPercentage uint64,
+) (*avalancheWarp.Message, bool, error) {
+	// Regardless of the response's relevance, call it's finished handler once this function returns
+	defer response.OnFinishedHandling()
+
+	// Check if this is an expected response.
+	m := response.Message()
+	rcvReqID, ok := message.GetRequestID(m)
+	if !ok {
+		// This should never occur, since inbound message validity is already checked by the inbound handler
+		s.logger.Error("Could not get requestID from message")
+		return nil, false, nil
```

This may have already been here, but we should return an error in this case.

---

This shouldn't happen but if it did for a single request, returning an error here instead of nil would short-circuit the entire aggregation. Do we still want to make the change?